### PR TITLE
add debug image to main pipeline

### PR DESCRIPTION
**Pull request by Claych**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by PettitWesley — APPROVED** [review]
✅

File: scripts/publish.sh [hunk 1]
```diff
@@ -434,8 +445,14 @@ publish_ecr() {
 		push_image_ecr ${AWS_ACCOUNT}.dkr.ecr.${AWS_REGION}.amazonaws.com/amazon/aws-for-fluent-bit-test:"$arch" \
 			${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit:"$arch"-${AWS_FOR_FLUENT_BIT_VERSION}
 
+		push_image_ecr ${AWS_ACCOUNT}.dkr.ecr.${AWS_REGION}.amazonaws.com/amazon/aws-for-fluent-bit-test:"$arch"-"debug" \
+			${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit:"$arch"-"debug"-${AWS_FOR_FLUENT_BIT_VERSION}
+
```
Comment (matthewfala) [inline scripts/publish.sh:450]: Am I missing it? Do we not have manifest lists for the debug image?

**Review by matthewfala — CHANGES_REQUESTED** [review]
Please comment on the manifest lists.

File: scripts/publish.sh (hunk 1 above)
Comment (Claych) [reply scripts/publish.sh:450]: agreed, will create manifest list for debug and init-debug images, thank you for your doc here

File: scripts/publish.sh [hunk 2]
```diff
@@ -424,6 +455,25 @@ create_manifest_list_init() {
...
+	# TODO: Add a way to automatically generate arch images in manifest
+	docker manifest create ${1}:${tag} ${1}:"$init"-arm64-debug-${version} ${1}:"$init"-amd64-debug-${version}
+
+	for arch in "${ARCHITECTURES[@]}"
+	do
+		docker manifest annotate --arch "$arch" ${1}:${tag} ${1}:"$init"-"$arch"-debug-${version}
+	done
+
+	# sanity check on the debug log.
+    docker manifest inspect ${1}:${tag}
```
Comment (matthewfala) [inline scripts/publish.sh]: strange formatting. Would you please fix?

File: scripts/publish.sh [hunk 3]
```diff
@@ -457,9 +507,13 @@ publish_ecr() {
...
 	create_manifest_list ${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit ${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
 	create_manifest_list ${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit "latest" ${AWS_FOR_FLUENT_BIT_VERSION}
+	create_manifest_list_debug ${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit "debug"-${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
+	create_manifest_list_debug ${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit "debug-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
 
 	create_manifest_list_init ${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit "$init"-${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
 	create_manifest_list_init ${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit "init-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
+	create_manifest_list_init_debug ${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit "$init"-"debug"-${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
+	create_manifest_list_init_debug ${account_id}.dkr.ecr.${region}.amazonaws.com/aws-for-fluent-bit "init-debug-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
```
Comment (matthewfala) [inline scripts/publish.sh]: should you use "$init"-debug-latest?
Not needed, but following convention is probably good idea.

File: scripts/publish.sh [hunk 4]
```diff
@@ -405,6 +417,25 @@ create_manifest_list() {
...
+    export DOCKER_CLI_EXPERIMENTAL=enabled
+    tag=${2}
+    version=${3}
+
+    # TODO: Add a way to automatically generate arch images in manifest
+    docker manifest create ${1}:${tag} ${1}:arm64-debug-${version} ${1}:amd64-debug-${version}
+
+    for arch in "${ARCHITECTURES[@]}"
+    do
+        docker manifest annotate --arch "$arch" ${1}:${tag} ${1}:"$arch"-debug-${version}
```
Comment (PettitWesley) [inline scripts/publish.sh]: nit: do we really need a new function here? You could pass to the function `debug-{version}` as the  third arg I think to the old function and it'd work.

File: scripts/publish.sh [hunk 5]
```diff
@@ -343,9 +351,13 @@ sync_latest_image() {
 	if [ "${account_id}" != "${classic_regions_account_id}" ]; then
 		create_manifest_list ${account_id}.dkr.ecr.${region}.${endpoint}/aws-for-fluent-bit "latest" ${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB}
 		create_manifest_list ${account_id}.dkr.ecr.${region}.${endpoint}/aws-for-fluent-bit ${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB} ${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB}
+		create_manifest_list_debug ${account_id}.dkr.ecr.${region}.${endpoint}/aws-for-fluent-bit "debug-latest" ${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB}
+		create_manifest_list_debug ${account_id}.dkr.ecr.${region}.${endpoint}/aws-for-fluent-bit "debug"-${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB} ${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB}
 
 		create_manifest_list_init ${account_id}.dkr.ecr.${region}.${endpoint}/aws-for-fluent-bit "init-latest" ${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB}
 		create_manifest_list_init ${account_id}.dkr.ecr.${region}.${endpoint}/aws-for-fluent-bit "$init"-${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB} ${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB}
+		create_manifest_list_init_debug ${account_id}.dkr.ecr.${region}.${endpoint}/aws-for-fluent-bit "init-debug-latest" ${AWS_FOR_FLUENT_BIT_VERSION_DOCKERHUB}
```
Comment (matthewfala) [inline scripts/publish.sh]: Should we use $init here? I'm fine if you don't. This would just be for consistancy

File: scripts/publish.sh [hunk 6]
```diff
@@ -201,10 +205,14 @@ publish_to_public_ecr() {
...
 		aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/aws-observability
 		create_manifest_list public.ecr.aws/aws-observability/aws-for-fluent-bit "latest" ${AWS_FOR_FLUENT_BIT_VERSION}
+		create_manifest_list_debug public.ecr.aws/aws-observability/aws-for-fluent-bit "debug-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
+		create_manifest_list_debug public.ecr.aws/aws-observability/aws-for-fluent-bit "debug"-${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
 
 		create_manifest_list_init public.ecr.aws/aws-observability/aws-for-fluent-bit "$init"-${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
 		aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/aws-observability
 		create_manifest_list_init public.ecr.aws/aws-observability/aws-for-fluent-bit "init-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
+		create_manifest_list_init_debug public.ecr.aws/aws-observability/aws-for-fluent-bit "init-debug-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
```
Comment (matthewfala) [inline scripts/publish.sh]: Should we use $init here? I'm fine if you don't. This would just be for consistancy

File: scripts/publish.sh [hunk 7]
```diff
@@ -160,9 +160,13 @@ publish_to_docker_hub() {
 
 		create_manifest_list ${1} "latest" ${AWS_FOR_FLUENT_BIT_VERSION}
 		create_manifest_list ${1} ${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
+		create_manifest_list_debug ${1} "debug-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
+		create_manifest_list_debug ${1} "debug"-${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
 
 		create_manifest_list_init ${1} "init-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
 		create_manifest_list_init ${1} "$init"-${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
+		create_manifest_list_init_debug ${1} "init-debug-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
```
Comment (matthewfala) [inline scripts/publish.sh]: Should we use $init here? I'm fine if you don't. This would just be for consistancy

File: scripts/publish.sh (hunk 6 above)
Comment (Claych) [reply scripts/publish.sh]: above this line, we used "init-latest" before, so I followed it with "init-debug-latest"

**Review by matthewfala — APPROVED** [review]
Just a minor comment on formatting. The other comments are optional. Looks good for the most part! Thank you Clay!

File: scripts/publish.sh (hunk 2 above)
Comment (Claych) [reply scripts/publish.sh]: this format is the same with create_manifest_list_init() function. I think this is the problem of github

File: scripts/publish.sh (hunk 4 above)
Comment (matthewfala) [reply scripts/publish.sh]: Yeah, this may be a good option.

File: scripts/publish.sh (hunk 3 above)
Comment (Claych) [reply scripts/publish.sh]: yes, maybe just follow convention

File: scripts/publish.sh (hunk 2 above)
Comment (matthewfala) [reply scripts/publish.sh]: It's because there are spaces rather than a tab.

Comment (matthewfala) [reply scripts/publish.sh]: Try delete the white space and use a tab instead like the rest of the whitespaces

Comment (Claych) [reply scripts/publish.sh]: yes, I think that's the point

File: scripts/publish.sh (hunk 4 above)
Comment (Claych) [reply scripts/publish.sh]: process in testing

File: scripts/publish.sh (hunk 7 above)
Comment (PettitWesley) [reply scripts/publish.sh]: I don't understand why we added $init in the first place... let's not increase its usage

File: scripts/publish.sh [hunk 8]
```diff
@@ -147,17 +147,26 @@ publish_to_docker_hub() {
...
 			docker tag ${1}:"$init"-"$arch" ${1}:"$init"-"${arch}"-${AWS_FOR_FLUENT_BIT_VERSION}
 			docker push ${1}:"$init"-"$arch"-${AWS_FOR_FLUENT_BIT_VERSION}
 
+			docker tag ${1}:"$init"-"$arch"-"debug" ${1}:"$init"-"${arch}"-"debug"-${AWS_FOR_FLUENT_BIT_VERSION}
+			docker push ${1}:"$init"-"$arch"-"debug"-${AWS_FOR_FLUENT_BIT_VERSION}
 		done
 
 		create_manifest_list ${1} "latest" ${AWS_FOR_FLUENT_BIT_VERSION}
 		create_manifest_list ${1} ${AWS_FOR_FLUENT_BIT_VERSION} ${AWS_FOR_FLUENT_BIT_VERSION}
+		create_manifest_list ${1} "debug-latest" debug-${AWS_FOR_FLUENT_BIT_VERSION}
```
Comment (PettitWesley) [inline scripts/publish.sh]: is the latest publishing rebased on top of my recently merged changes to support non-latest releases?

Comment (Claych) [reply scripts/publish.sh]: no.. will update to meet the changes

File: scripts/publish.sh [hunk 9]
```diff
@@ -209,22 +219,36 @@ publish_to_public_ecr() {
...
 		if [ "${PUBLISH_LATEST}" = "true" ]; then
 			create_manifest_list public.ecr.aws/aws-observability/aws-for-fluent-bit "latest" ${AWS_FOR_FLUENT_BIT_VERSION}
 			aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/aws-observability
+			create_manifest_list public.ecr.aws/aws-observability/aws-for-fluent-bit "debug-latest" debug-${AWS_FOR_FLUENT_BIT_VERSION}
+			aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/aws-observability
 
 			create_manifest_list_init public.ecr.aws/aws-observability/aws-for-fluent-bit "init-latest" ${AWS_FOR_FLUENT_BIT_VERSION}
 			aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/aws-observability
+			create_manifest_list_init public.ecr.aws/aws-observability/aws-for-fluent-bit "init-debug-latest" debug-${AWS_FOR_FLUENT_BIT_VERSION}
+			aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/aws-observability
```
Comment (PettitWesley) [inline scripts/publish.sh]: I know my changes had this, but log in again isn't needed as final command in this function I think, just before we create any manifest

**Review by PettitWesley — APPROVED** [review]

**Review by matthewfala — APPROVED** [review]
Looks good!

File: scripts/publish.sh (hunk 9 above)
Comment (Claych) [reply scripts/publish.sh]: got it

**Comment by PettitWesley** [comment]
@Claych Let's include this in our next release. Can you please add an entry in the debugging.md and README.md for the new images?

File: README.md [hunk 10]
```diff
@@ -181,6 +181,10 @@ For Windows, [we release images](#windows-images) only for amd64 CPU architectur
 - Windows Server 2019
 - Windows Server 2022
 
+#### debug images
```
Comment (PettitWesley) [inline README.md]: needs a table of contents link

File: README.md [hunk 11]
```diff
@@ -181,6 +181,10 @@ For Windows, [we release images](#windows-images) only for amd64 CPU architectur
...
+
+The Fluent Bit Debug Image is a containerized environment designed for testing and debugging Fluent Bit configurations, and it also has the capability to transfer core dump files to an S3 bucket for further analysis and troubleshooting. This feature can help streamline the debugging process and facilitate quicker resolution of issues.
```
Comment (PettitWesley) [inline README.md]: give the tag format for the debug images

Comment (PettitWesley) [reply README.md]: Also I'd prefer to rephrase this as:

> Deploying AWS for Fluent debug images can help the AWS team troubleshoot an issue. If you experience a bug, especially a crash/SIGSEGV issue (link to debugging guide SIGSEGV section), then please consider deploying the debug version of the image. After a crash, the debug image can print out a stacktrace and upload a core dump to S3. See our debugging guide {link} for more info on using debug images.

File: troubleshooting/debugging.md [hunk 12]
```diff
@@ -816,7 +816,7 @@ When you recieve a SIGSEGV/crash report from a FireLens customer, perform the fo
 
 ##### 1. Build and distribute a core dump S3 uploader image
 
-You need a customized image build for the specific version/case you are testing. Make sure the `ENV FLB_VERSION` is set to the right version in the `Dockerfile.debug-base` and make sure the `AWS_FLB_CHERRY_PICKS` file has the right contents for the release you are testing. 
+You can find and download Fluent Bit Debug Images for specific versions on Docker Hub and public ECR repositories. If you need a customized image build for the specific version/case you are testing. Make sure the `ENV FLB_VERSION` is set to the right version in the `Dockerfile.debug-base` and make sure the `AWS_FLB_CHERRY_PICKS` file has the right contents for the release you are testing.
```
Comment (PettitWesley) [inline troubleshooting/debugging.md]: note the tag names again

File: README.md (hunk 11 above)
Comment (PettitWesley) [reply README.md]: We possibly should explicitly say not to deploy these to prod... I also still want to run the load tests on them though actually and call that out... I really think it should be safe to deploy these in prod...

File: README.md [hunk 13]
```diff
@@ -181,6 +182,10 @@ For Windows, [we release images](#windows-images) only for amd64 CPU architectur
 - Windows Server 2019
 - Windows Server 2022
 
+#### Debug images
+
+Deploying AWS for Fluent `debug` images can help the AWS team troubleshoot an issue. If you experience a bug, especially a [crash/SIGSEGV issue](https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#segfaults-and-crashes-sigsegv), then please consider deploying the debug version of the image. After a crash, the debug image can print out a stacktrace and upload a core dump to S3. See our [debugging guide](https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#guide-to-debugging-fluent-bit-issues) for more info on using debug images.
```
Comment (PettitWesley) [inline README.md]: 1st link should go here: https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#caught-signal-sigsegv

2nd should go here: https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#1-build-and-distribute-a-core-dump-s3-uploader-image

File: README.md [hunk 14]
```diff
@@ -14,6 +14,7 @@
 - [Use Case Guide](use_cases/)
 - [Public Images](#public-images)
     - [arm64 and x86 images](#arm64-and-amd64-images)
+    - [Debug images](#debug-images)
```
Comment (PettitWesley) [inline README.md]: I think this should be the last entry in this section

File: README.md (hunk 13 above)
Comment (PettitWesley) [reply README.md]: Please also give the tag format, if I want debug image for 2.X.Y, what tag is it?

**Review by PettitWesley — CHANGES_REQUESTED** [review]
